src/lib/xof/xof.h

```diff
@@ -0,0 +1,196 @@
+       * @param salt  a salt value to parameterize the XOF
+       * @param key   a key to parameterize the XOF
+       */
+      virtual void start(std::span<const uint8_t> salt = {}, std::span<const uint8_t> key = {}) = 0;
+
+      /**
+       * @returns true if salt length is acceptable, false otherwise
+       */
+      virtual bool valid_salt_length(size_t salt_len) = 0;
+
+      /**
+       * @returns an object describing limits on the key size
+       */
+      virtual Key_Length_Specification key_spec() const = 0;
+
+      /**
+       * @return the intrinsic processing block size of this XOF
+       */
+      virtual size_t block_size() const { return 0; }
+
+      /**
+       * Return a new XOF object with the same state as *this.
+       *
+       * If the XOF is not yet in the output phase, it efficiently allows
+       * using several messages with a common prefix.
+       * Otherwise, the copied state will produce the same output
+       * bit stream as the original object at the time of this invocation.
+       *
+       * This function should be called `clone` but for consistency with
+       * other classes it is called `copy_state`.
+       *
+       * @return new XOF object
+       */
+      virtual std::unique_ptr<XOF> copy_state() const = 0;
+
+      /**
+       * @return new object representing the same algorithm as *this
+       */
+      virtual std::unique_ptr<XOF> new_object() const = 0;
+
+      /**
+       * Typically, this is `true` for new objects and becomes `false`
+       * once `output()` was called for the first time.
+       *
+       * @returns true iff calling `update()` is legal in the current object state
+       */
+      virtual bool accepts_input() const = 0;
+
+      /**
+       * Add @p input data to the XOF's internal state
+       *
+       * @param input  the data that shall be
+       */
+      void update(std::span<const uint8_t> input) { add_data(input); }
+
+      /**
+       * @return the next @p bytes output bytes as the specified container type @p T.
+       */
+      template <concepts::resizable_byte_buffer T = secure_vector<uint8_t>>
+      T output(size_t bytes) {
+         T out(bytes);
+         generate_bytes(out);
+         return out;
+      }
+
+      /**
+       * Convenience overload to generate a std::vector<uint8_t>. Same as calling
+       * `XOF::output<std::vector<uint8_t>>()`.
+       *
+       * @return the next @p bytes output bytes as a byte vector.
+       */
+      std::vector<uint8_t> output_stdvec(size_t bytes) { return output<std::vector<uint8_t>>(bytes); }
+
+      /**
+       * Fill @p output with the next output bytes. The number of bytes
+       * depends on the size of @p output.
+       */
+      void output(std::span<uint8_t> output) { generate_bytes(output); }
+
+      /**
+       * @return the next single output byte
+       */
+      uint8_t output_next_byte() {
+         uint8_t out;
+         generate_bytes({&out, 1});
+         return out;
+      }
```

Comment:
`XOF` does not inherit from a generic `Buffered_Input` class anymore. Instead it directly defines `update(std::span<>)` and no additional convenience methods. For consistency, we should probably remove the convenience methods of `output()` as well, no?

My two cents: I did enjoy having some convenience overloads (particularly taking `std::string_view` and Big/Little-Endian integers). Re-implementing them here as well as later in `HashFunction` and `MessageAuthenticationCode` feels like violating the DRY principle, though. Hence, I wouldn't do that.

Nevertheless, I'd suggest keeping two flavors of `output()`:

* `void output(std::span<>)`
   to allow the user to handle memory allocation
* `T output(size_t bytes)` with `T` being a user-defined container that defaults to `secure_vector<>`
   to let the user conveniently pull data wrapped in their favorite vector-ish.